### PR TITLE
Extract performance improvement chapter

### PR DESCRIPTION
**Pull request by malandrina**

### CODE REVIEW

File: book/improving_application_performance/geocode_only_when_necessary.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
+## Geocode Only When Necessary
+
+Currently we're geocoding `Location` objects every time they're validated. This
```
Comment (joshuaclayton) [inline book/improving_application_performance/geocode_only_when_necessary.md]: "every time they're validated" => "in an `after_validation` callback"?

File: book/improving_application_performance/geocode_only_when_necessary.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+approach is less than ideal because it makes our application more likely to hit
+the daily rate limit of the external geocoding service. In addition, we're
+slowing down our application with unnecessary external requests: Geocoding with
+Google takes an average of 75ms. An easy way to improve improve application
+performance is to only geocode when necessary.
```
Comment (joshuaclayton) [inline book/improving_application_performance/geocode_only_when_necessary.md]: "to only geocode" => "to geocode only"

Comment (joshuaclayton) [inline book/improving_application_performance/geocode_only_when_necessary.md]: "when necessary" feels pretty arbitrary; can we explain this a bit more? I'm not sure what would sound concise, though - "when the address changes" seems most applicable for now. The branch https://github.com/thoughtbot/geocoding_on_rails/tree/lw-use-geocoder-stubs-in-integration-tests adds additional logic to skip geocoding a new record if lat/lng are provided but I think we can skip that here for now. Thoughts?

File: book/improving_application_performance/geocode_only_when_necessary.md [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+First, we add a condition to the `after_validation` callback on the `Location`
+model:
+
+` app/models/location.rb@c7275e2:1,8
+
+Then we build out the `#address_changed?` method to ensure that an object is
```
Comment (joshuaclayton) [inline book/improving_application_performance/geocode_only_when_necessary.md]: "Then" => "Second"?

Comment (joshuaclayton) [inline book/improving_application_performance/geocode_only_when_necessary.md]: "to ensure that an object" => "to ensure the location instance"?

**Comment by joshuaclayton** [comment]
@malandrina I glossed over the fact that this is is just moving the performance improvement text to separate files. These changes don't necessarily need to be made for this change, just didn't want to lose my thoughts here.

File: book/improving_application_performance/geocode_only_when_necessary.md (hunk 2 above)
Comment (malandrina) [inline book/improving_application_performance/geocode_only_when_necessary.md]: I agree! "when necessary" is unclear. I'm changing it to "when the address changes".

**Comment by malandrina** [comment]
@joshuaclayton I think your suggestions are good. I made the changes you suggested in https://github.com/thoughtbot/geocoding_on_rails/commit/d588e0b28eb3141aa3b800d56def591873e8f299.

**Comment by joshuaclayton** [comment]
@malandrina one last comment - aside from that, this looks great! Merge whenever you're ready!

**Pull request merged**